src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -314,11 +322,21 @@ protected void ClearSolution()
         /// Call the base method at the end of your method.
         /// </summary>
         protected virtual void ClearSolutionData()
+        {
+            using (_serializationLock.DisposableWait())
+            {
+                ClearSolutionData_NoLock();
+            }
+        }
+
+        private (Solution oldSolution, Solution newSolution) ClearSolutionData_NoLock()
```

Comment:
Our pattern in other places for _NoLock methods is to stick an assert asserting the semaphore's count is zero, something like:

https://github.com/dotnet/roslyn/blob/c99037b63105f997e9d0d92bb6a9a6d227834ab6/src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.cs#L266

Obviously this isn't a _perfect_ assert -- if we forgot to lock it another thread could still be holding the lock at the time of the assert, but the hope is in a single-threaded unit test you won't get lucky that way.

Especially for this type it'd make it very clear _which_ lock needs to be held, because there's multiple.

Reply:
Thanks for changing -- looks like we should also just make an extension here because clearly our decisions on how to do the check are opposite. :-)